frontend/catalyst/compilation_pipelines.py

```diff
@@ -511,14 +524,18 @@ def __init__(self, fn, compile_options):
         preferred_workspace_dir = (
             pathlib.Path.cwd() if self.compile_options.keep_intermediate else None
         )
+        self.preferred_workspace_dir = preferred_workspace_dir
+
         # If we are compiling from textual ir, just use this as the name of the function.
         name = "compiled_function"
         if not self.compiling_from_textual_ir:
             # pylint: disable=no-member
             # Guaranteed to exist after functools.update_wrapper AND not compiling from textual IR
             name = self.__name__
+        self.function_name = name
 
-        self.workspace = WorkspaceManager.get_or_create_workspace(name, preferred_workspace_dir)
+        if not self.compile_options.static_argnums:
+            self.workspace = WorkspaceManager.get_or_create_workspace(name, preferred_workspace_dir)
```

Comment:
There is another workspace creation in ``__call__`` if ``self.compile_options.static_argnums`` is not empty. Or maybe I should create a new workspace after compiling a function instead?

Reply:
Sounds good, so the workspace is set when compilation actually happens, and each compiled function uses its own temporary workspace.

Do you think there would be any issues with the workspace object being garbage collected? (previously it was tied to the lifetime of the qjit object)

Reply:
Oh never mind, nice refactor 💯

Reply:
I think it is unsafe right now since we do not keep the references of old workspace objects. I will modify it to store references of workspace objects.

Reply:
Yes I think this boils down to the question of whether we want to close and reopen a shared library when it is no longer needed, or if we are okay having all versions of a function loaded in memory at the same time.

In the latter case, hanging on the workspace isn't needed, but if we want the `SharedObjectManager` to reopen a previously closed library we would need to keep the binary file around.

Reply:
I guess there might be some scenarios with large and complex functions. Maybe "close and reopen" is preferred?

Reply:
I made a way to close and reopen a shared object instead of storing a function in memory. Commit 
[3ecf7a9](url=https://github.com/PennyLaneAI/catalyst/pull/476/commits/3ecf7a966da07e1ab9021f1f3dbd065dd8db9b81) made the changes. Before this commit, we stored all functions in memory. Please let me know which makes more sense.

Reply:
Hi @tzunghanjuang, I don't see how the commit implements closing and reopening 🤔 From what I can tell it just bypasses the closing of the shared library, as in your previous version, but in a different place (the `SharedObjectManager`). Please correct me if I'm wrong :)

Looking at this change in particular:
```py
        if delete_files:
            dlclose = ctypes.CDLL(None).dlclose
            dlclose.argtypes = [ctypes.c_void_p]
            # pylint: disable=protected-access
            dlclose(self.shared_object._handle)
```
I would say that the flag is a misnomer, since it doesn't interact with the file system, but rather determines whether the shared library is actually closed (with `dlclose`). Additionally, there is now an execution path in the `close` function that doesn't actually close the library, which is confusing.

Might I suggest an implementation of the open/close strategy where:
- the library is closed with `dlclose` upon recompilation (as before)
- encountering a known hash reopens the shared library from disk by accessing the temporary workspace
  - note that the `SharedObjectManager` doesn't actually have a way to re-open the lib, but this can easily be fixed by factoring out the open code from the constructor
- maybe it would make sense to tie the workspace lifetime to that of a `CompiledFunction`?
  - a `SharedObjectManager` is also alive for as long as the `CompiledFunction` is
  - the compiled function cache stores instances of `CompiledFunction`, so it would make sense that each `CompiledFunction` also manages its workspace -> switching the "active" compiled function would then also switch the workspace and shared lib manager